scripts/fit_pgd.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+from absl import app
+from absl import flags
+import matplotlib.pyplot as plt
+import tensorflow as tf
+import pickle
+import flowpm
+from flowpm import tfpm
+from flowpm.utils import cic_paint
+from flowpm.tfpower import linear_matter_power
+from functools import partial
+import flowpm.scipy.interpolate as interpolate
+
+from nbodykit.cosmology import Cosmology
+from nbodykit.cosmology.power.halofit import HalofitPower
+from astropy.cosmology import Planck15
+import astropy.units as u
+
+flags.DEFINE_string("filename", "results_fit_PGD.pkl", "Output filename")
+flags.DEFINE_integer("niter", 51, "Number of iterations of the first PGD fit")
+flags.DEFINE_integer("niter_refine", 11,
+                     "Number of iterations of subsequent PGD fit")
+flags.DEFINE_float("learning_rate", 0.1, "ADAM learning rate for the PGD optim")
+flags.DEFINE_integer("batch_size", 8, "Number of random draws")
+
+flags.DEFINE_float("alpha0", 0.3, "Initial guess for alpha at z=0")
+flags.DEFINE_float("kl0", 0.3, "Initial guess for kl at z=0")
+flags.DEFINE_float("ks0", 10, "Initial guess for ks at z=0")
+
+flags.DEFINE_boolean("fix_scales", False,
+                     "Whether to fix the scales after the initial fit at z=0")
+
+flags.DEFINE_float("a_init", 0.1, "Initial scale factor")
+flags.DEFINE_integer("nsteps", 40, "Number of steps in the N-body simulation")
+flags.DEFINE_float("box_size", 64.,
+                   "Transverse comoving size of the simulation volume")
+flags.DEFINE_integer("nc", 64,
+                     "Number of transverse voxels in the simulation volume")
+flags.DEFINE_integer("B", 1, "Scale resolution factor")
+
+FLAGS = flags.FLAGS
+
+
+def pgd_loss(alpha, scales, state, target_pk, return_pk=False):
+  """
+  Defines the loss function for the PGD parameters
+  """
+  shape = state.get_shape()
+  batch_size = shape[1]
+  pgdparams = tf.concat([alpha, scales], 0)
+
+  # Step I: Apply PGD to the state vector
+  pdgized_state = state[0] + tfpm.PGD_correction(
+      state, pgdparams, nc=[FLAGS.nc] * 3, pm_nc_factor=FLAGS.B)
+
+  # Step II: Painting
+  field = cic_paint(tf.zeros([batch_size] + [FLAGS.nc] * 3), pdgized_state)
+
+  # Step III: Compute power spectrum
+  k, pk = flowpm.power_spectrum(
+      field,
+      boxsize=np.array([FLAGS.box_size] * 3),
+      kmin=np.pi / FLAGS.box_size,
+      dk=2 * np.pi / FLAGS.box_size)
+  # Averaging pk over realisations
+  pk = tf.reduce_mean(pk, axis=0)
+
+  # Step IV: compute loss
+  weight = 1 - k / (np.pi * FLAGS.nc / FLAGS.box_size)
```

Comment:
In my experiments, I could not settle on the weight, unless you did. Note that this is not the weight that Biwei uses in his paper, it was something I came up with, adhoc. 
So we might want to keep this highlighted somehow so that we remember what is being done. Maybe use an argument (FLAGS) to the script instead of just keeping it under the wraps here.

Reply:
I have spent exactly 0 minutes tuning the weights..... are you saying this is a really bad choice? why do we need a weight anyway actually?

Reply:
I am not saying this is a bad choice. I think this one worked for me. But this is adhoc and we should be aware that it is there.

The reason I started including weights is because there are competing effects that we do not want to correct with PGD - on large scales, your PS will not agree with theory spectra due to cosmic variance; on small scales, we have the CIC window kicking in for which we are not correctly compensating  (are we?). weights were a way to downweight the loss function on these scales which are dominated by these effects so that PGD can focus on what we actually care about. 
Incidentally...fitting for a high resolution simulation instead of halofit will automatically take care of these effects ;)

Reply:
ok. well... so this is not great, and we should really use the cic painting compensation on small scales. On large scales, in the current script we are averaging the power spectrum over several realisations, so that should take care of cosmic variance more or less.

For now, I'm following your suggestion, I've made your custom weight a parameter. We will open another issue to improve this further.